source/grid/manifold.cc

```diff
@@ -157,85 +157,54 @@ normal_vector (const Triangulation<3, 3>::face_iterator &face,
-          t2 = get_tangent_vector(p, face->vertex(2));
-        }
-      else
+      // if p is too close to vertex i try again with different i and j
+      if ((p - face->vertex(i)).norm_square() <
+          std::numeric_limits<double>::epsilon() * (p - face->vertex(j)).norm_square())
         {
-          t1 = get_tangent_vector(p, face->vertex(0));
-          t2 = get_tangent_vector(p, face->vertex(1));
-        }
-    }
-  else
-    {
-      // we are closer to one of the vertices than to the
-      // center of the face
-      switch (min_index)
-        {
-        case 0:
-        {
-          t1 = get_tangent_vector(p, face->vertex(1));
-          t2 = get_tangent_vector(p, face->vertex(2));
-          break;
-        }
-        case 1:
-        {
-          t1 = get_tangent_vector(p, face->vertex(3));
-          t2 = get_tangent_vector(p, face->vertex(0));
-          break;
-        }
-        case 2:
-        {
-          t1 = get_tangent_vector(p, face->vertex(0));
-          t2 = get_tangent_vector(p, face->vertex(3));
-          break;
-        }
-        case 3:
-        {
-          t1 = get_tangent_vector(p, face->vertex(2));
-          t2 = get_tangent_vector(p, face->vertex(1));
-          break;
-        }
-        default:
-          Assert(false, ExcInternalError());
-          break;
+          ++i;
+          continue;
         }
+
+      // if p is too close to vertex j try again with different j
+      if ((p - face->vertex(j)).norm_square() <
+          std::numeric_limits<double>::epsilon() * (p - face->vertex(i)).norm_square())
+        continue;
+
+      t1 = get_tangent_vector(p, face->vertex(i));
+      t2 = get_tangent_vector(p, face->vertex(j));
+
+      normal = cross_product_3d(t1,t2);
+
+      // if t1 and t2 are (nearly) linearly dependent try again with different j / t2
+      if (normal.norm_square() < std::numeric_limits<double>::epsilon() *
+          t1.norm_square() * t2.norm_square())
+        continue;
+
+      break;
     }
 
-  const Tensor<1,spacedim> normal = cross_product_3d(t1,t2);
+  Assert(normal.norm_square() >= std::numeric_limits<double>::epsilon() *
+         t1.norm_square() * t2.norm_square(),
+         ExcMessage("Manifold::normal_vector was unable to find a suitable combination "
+                    "of vertices to compute a normal on this face. Check for distorted "
+                    "faces in your triangulation."));
+
+  // Make sure all found normal vectors on this face point in the same direction
+  // as the 'reference' normal vector created at the center position.
+  const Point<spacedim> center = face->center();
+  const Tensor<1,spacedim> reference_t1 = get_tangent_vector(center, face->vertex(0));
+  const Tensor<1,spacedim> reference_t2 = get_tangent_vector(center, face->vertex(1));
+  const Tensor<1,spacedim> reference_normal = cross_product_3d(reference_t1,reference_t2);
+
+  if (reference_normal * normal < 0.0)
+    normal *= -1;
```

Comment:
Can you explain why you need this? Is it because you've lost track of whether vertices `i` and `j` are oriented in clockwise orientation with an angle of less than 180 degrees when seen from point `p`? That's when you get the sign flip. But I wonder whether you can detect that case without having to call `get_tangent_vector` twice as that might be an expensive operation.